release/train_tests/benchmark/dataloader_factory.py

```diff
@@ -1,13 +1,498 @@
+
+    This dataset:
+    1. Reads Parquet files from S3 one row group at a time
+    2. Processes images with optional random transforms
+    3. Yields (image, label) tensors
+    4. Supports row limits per worker for controlled data processing
+    """
+
+    LOG_FREQUENCY = 1000  # Log progress every 1000 rows
+
+    def __init__(
+        self,
+        file_urls: List[str],
+        random_transforms: bool = True,
+        limit_rows_per_worker: Optional[int] = None,
+    ):
+        """Initialize the dataset.
+
+        Args:
+            file_urls: List of S3 URLs to load
+            random_transforms: Whether to use random transforms for training
+            limit_rows_per_worker: Maximum number of rows to process per worker (None for all rows)
+        """
+        S3Reader.__init__(self)
+        self.file_urls = file_urls
+        self.limit_rows_per_worker = limit_rows_per_worker
+        self.random_transforms = random_transforms
+
+    def _read_parquet_file(self, file_url: str) -> Iterator[pd.DataFrame]:
+        """Read a Parquet file from S3 one row group at a time.
+
+        Args:
+            file_url: S3 URL of the Parquet file
+
+        Yields:
+            DataFrame containing one row group at a time
+        """
+        try:
+            start_time = time.time()
+            logger.info(f"[Dataset] Reading Parquet file: {file_url}")
+
+            # Get parquet file metadata
+            bucket, key = self._parse_s3_url(file_url)
+            response = self.s3_client.get_object(Bucket=bucket, Key=key)
+            parquet_file = pq.ParquetFile(io.BytesIO(response["Body"].read()))
+            num_row_groups = parquet_file.num_row_groups
+
+            logger.info(f"[Dataset] Found {num_row_groups} row groups in {file_url}")
+
+            for row_group in range(num_row_groups):
+                # Read row group and convert to pandas
+                table = parquet_file.read_row_group(row_group)
+                df = table.to_pandas()
+                yield df
+
+            total_time = time.time() - start_time
+            logger.info(f"[Dataset] Completed reading {file_url} in {total_time:.2f}s")
+
+        except Exception as e:
+            logger.error(f"[Dataset] Error reading file {file_url}: {str(e)}")
+            raise
+
+    def __iter__(self) -> Iterator[Tuple[torch.Tensor, torch.Tensor]]:
+        """Main iteration method that processes files and yields (image, label) tensors.
+
+        This method:
+        1. Distributes files among workers
+        2. Processes rows with image transforms
+        3. Converts to tensors
+        4. Respects row limits per worker
+        """
+        try:
+            # Get worker info for file distribution
+            worker_info = torch.utils.data.get_worker_info()
+            worker_id = worker_info.id if worker_info else 0
+            num_workers = worker_info.num_workers if worker_info else 1
```

Comment:
Just to clarify, these are dataloader workers, not training workers?

Do we still need to shard urls at the training worker level?

Reply:
i.e. with `ray.train.get_context().get_world_rank()`